source/common/filter/filter_config_discovery_impl.cc

```diff
@@ -0,0 +1,197 @@
+              "{}). Ignoring.",
+              removed_resources[0]);
+  }
+  if (!added_resources.empty()) {
+    onConfigUpdate(added_resources, added_resources[0].get().version());
+  }
+}
+
+void HttpFilterConfigSubscription::onConfigUpdateFailed(Config::ConfigUpdateFailureReason reason,
+                                                        const EnvoyException*) {
+  ENVOY_LOG(debug, "Updating filter config {} failed due to {}", filter_config_name_, reason);
+  stats_.config_fail_.inc();
+  // Make sure to make progress in case the control plane is temporarily failing.
+  init_target_.ready();
+}
+
+HttpFilterConfigSubscription::~HttpFilterConfigSubscription() {
+  // If we get destroyed during initialization, make sure we signal that we "initialized".
+  init_target_.ready();
+  // Remove the subscription from the provider manager.
+  filter_config_provider_manager_.subscriptions_.erase(subscription_id_);
+}
+
+std::shared_ptr<HttpFilterConfigSubscription> HttpFilterConfigProviderManagerImpl::getSubscription(
+    const envoy::config::core::v3::ConfigSource& config_source, const std::string& name,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix) {
+  // HttpFilterConfigSubscriptions are unique based on their config source and filter config name
+  // combination.
+  // TODO(https://github.com/envoyproxy/envoy/issues/11967) Hash collision can cause subscription aliasing.
+  const std::string subscription_id = absl::StrCat(MessageUtil::hash(config_source), ".", name);
+  auto it = subscriptions_.find(subscription_id);
+  if (it == subscriptions_.end()) {
+    auto subscription = std::make_shared<HttpFilterConfigSubscription>(
+        config_source, name, factory_context, stat_prefix, *this, subscription_id);
+    subscriptions_.insert(
+        {subscription_id, std::weak_ptr<HttpFilterConfigSubscription>(subscription)});
+    return subscription;
+  } else {
+    auto existing = it->second.lock();
+    RELEASE_ASSERT(existing != nullptr,
+                   absl::StrCat("Cannot find subscribed filter config resource ", name));
+    return existing;
+  }
+}
+
+HttpFilterConfigProviderPtr HttpFilterConfigProviderManagerImpl::createDynamicFilterConfigProvider(
+    const envoy::config::core::v3::ConfigSource& config_source,
+    const std::string& filter_config_name, bool require_terminal,
+    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
+    bool apply_without_warming) {
+  auto subscription =
+      getSubscription(config_source, filter_config_name, factory_context, stat_prefix);
+  // For warming, wait until the subscription receives the first response to indicate readiness.
+  // Otherwise, mark ready immediately and start the subscription on initialization. A default
+  // config is expected in the latter case.
+  if (!apply_without_warming) {
+    factory_context.initManager().add(subscription->initTarget());
```

Comment:
Is this server config factory context accurately modeling the listener warming chain? I.e. is this going to ensure listener warming.

Reply:
This is the context from HCM which I assume is listener context. I think RDS uses the init manager from this context. Let me know if I misunderstand something.